### PR TITLE
CUAEV refactor to initialize aev parameter only once

### PR DESCRIPTION
**Pull request by yueyericardo**

### CODE REVIEW

**Comment by yueyericardo** [comment]
original result: 
```
RUN                        Total AEV    Forward      Backward     Force        Optimizer    Others       Epoch time   GPU
0 cu Energy train          21.9 ms      13.1 ms      15.5 ms      0.0 ms       11.0 ms      3.6 ms       65.2 ms      1315.5MB
1 py Energy train          80.7 ms      12.9 ms      16.5 ms      0.0 ms       13.0 ms      3.7 ms       126.8 ms     2123.5MB
2 cu Energy + Force train  21.7 ms      12.9 ms      131.8 ms     75.2 ms      6.7 ms       5.0 ms       253.3 ms     2761.5MB
3 py Energy + Force train  80.2 ms      12.9 ms      435.0 ms     167.4 ms     6.9 ms       5.4 ms       707.8 ms     5667.5MB
```
original result (only forward):

```
File: small.pdb, Molecule size: 264

Original TorchANI:
  Duration: 0.42 s
  Speed: 2.11 ms/it

CUaev:
  Duration: 0.08 s
  Speed: 0.40 ms/it
  Speed up: 5.24 X

----------------------------------------------------------------------

File: 1hz5.pdb, Molecule size: 973

/home/richard/dev/torchani/torchani/resources/
Original TorchANI:
  Duration: 0.54 s
  Speed: 2.70 ms/it

CUaev:
  Duration: 0.19 s
  Speed: 0.93 ms/it
  Speed up: 2.92 X

----------------------------------------------------------------------

File: 6W8H.pdb, Molecule size: 3410

/home/richard/dev/torchani/torchani/resources/
Original TorchANI:
  Duration: 1.37 s
  Speed: 6.83 ms/it

CUaev:
  Duration: 0.94 s
  Speed: 4.68 ms/it
  Speed up: 1.46 X

----------------------------------------------------------------------
```

original result (forward, backward, doublebackward):

```
File: small.pdb, Molecule size: 264

Original TorchANI:
  Duration: 3.19 s
  Speed: 15.95 ms/it

CUaev:
  Duration: 0.49 s
  Speed: 2.46 ms/it
  Speed up: 6.49 X

----------------------------------------------------------------------

File: 1hz5.pdb, Molecule size: 973

/home/richard/dev/torchani/torchani/resources/
Original TorchANI:
  Duration: 3.63 s
  Speed: 18.14 ms/it

CUaev:
  Duration: 0.70 s
  Speed: 3.49 ms/it
  Speed up: 5.19 X

----------------------------------------------------------------------

File: 6W8H.pdb, Molecule size: 3410

/home/richard/dev/torchani/torchani/resources/
Original TorchANI:
  Duration: 4.62 s
  Speed: 23.10 ms/it

CUaev:
  Duration: 1.46 s
  Speed: 7.32 ms/it
  Speed up: 3.16 X

----------------------------------------------------------------------
```

File: torchani/cuaev/cuaev.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+CuaevComputer::CuaevComputer(
+    double Rcr,
+    double Rca,
+    const Tensor& EtaR_t,
+    const Tensor& ShfR_t,
+    const Tensor& EtaA_t,
+    const Tensor& Zeta_t,
+    const Tensor& ShfA_t,
+    const Tensor& ShfZ_t,
+    int64_t num_species) {
```
Comment (zasdfgbnm) [inline torchani/cuaev/cuaev.cpp:46]: Please use initializer list https://en.cppreference.com/w/cpp/language/constructor

File: torchani/cuaev/aev.h [hunk 2]
```diff
@@ -191,9 +195,20 @@ struct CuaevComputer : torch::CustomClassHolder {
       const Tensor& ShfZ_t,
       int64_t num_species);
 
-  Result forward(const Tensor& coordinates_t, const Tensor& species_t);
-  Tensor backward(const Tensor& grad_e_aev, const torch::intrusive_ptr<Result>& res_pt);
-  Tensor double_backward(const Tensor& grad_force, const torch::intrusive_ptr<Result>& res_pt);
+  Result forward(const Tensor& coordinates_t, const Tensor& species_t) {
+    Result result = cuaev_forward(coordinates_t, species_t, aev_params);
+    return result;
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.h:200]:
```suggestion
    return cuaev_forward(coordinates_t, species_t, aev_params);
```

File: torchani/cuaev/aev.h [hunk 3]
```diff
@@ -191,9 +195,20 @@ struct CuaevComputer : torch::CustomClassHolder {
...
+  }
+
+  Tensor backward(const Tensor& grad_e_aev, const Result& result) {
+    Tensor force = cuaev_backward(grad_e_aev, aev_params, result);
+    return force;
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.h:205]:
```suggestion
    return cuaev_backward(grad_e_aev, aev_params, result);
```

File: torchani/cuaev/aev.h [hunk 4]
```diff
@@ -191,9 +195,20 @@ struct CuaevComputer : torch::CustomClassHolder {
...
+  }
+
+  Tensor double_backward(const Tensor& grad_force, const Result& result) {
+    Tensor grad_grad_aev = cuaev_double_backward(grad_force, aev_params, result);
+    return grad_grad_aev;
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.h:209]:
```suggestion
    return cuaev_double_backward(grad_force, aev_params, result);
```

File: torchani/cuaev/aev.h [hunk 5]
```diff
@@ -0,0 +1,232 @@
+#ifndef CUAEV_COMPUTER
+#define CUAEV_COMPUTER
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.h:2]:
```suggestion
#pragma once
```

File: torchani/cuaev/aev.h [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+class CuaevAutograd : public torch::autograd::Function<CuaevAutograd> {
+ public:
+  static Tensor forward(
+      AutogradContext* ctx,
+      const Tensor& coordinates_t,
+      const Tensor& species_t,
+      const torch::intrusive_ptr<CuaevComputer>& cuaev_computer);
+  static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs);
+};
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.h:229]: Should we move them entirely to `cuaev.cpp`? I don't think it is shared between aev.cu and cuaev.cpp, so it does not need to be in header

File: torchani/cuaev/aev.h [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+#endif
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.h:231]: remove

File: torchani/cuaev/aev.h (hunk 6 above)
Comment (yueyericardo) [reply torchani/cuaev/aev.h:229]: you are right, could this speed up compiling cu file?

File: torchani/cuaev/cuaev.cpp [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+      tensor_numPairsPerCenterAtom(tensors[8]),
+      tensor_centerAtomStartIdx(tensors[9]),
+      maxnbrs_per_atom_aligned(tensors[10].item<int>()),
+      angular_length_aligned(tensors[11].item<int>()),
+      ncenter_atoms(tensors[12].item<int>()),
+      coordinates_t(tensors[13]),
+      species_t(tensors[14]) {}
+
+tensor_list Result::to_list() {
+  return {aev_t,
```
Comment (zasdfgbnm) [inline torchani/cuaev/cuaev.cpp:82]: should we remove aev from here, and in constructor `Result::Result(tensor_list tensors)`, just initialize `aev_t` to undefined tensor?

File: torchani/cuaev/cuaev.cpp [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    int64_t num_species)
+    : aev_params(Rcr, Rca, EtaR_t, ShfR_t, EtaA_t, Zeta_t, ShfA_t, ShfZ_t, num_species) {}
+
+Tensor CuaevDoubleAutograd::forward(
+    AutogradContext* ctx,
+    Tensor grad_e_aev,
+    const torch::intrusive_ptr<CuaevComputer>& cuaev_computer,
+    tensor_list result_tensors) {
+  Result result(result_tensors);
+  Tensor grad_coord = cuaev_computer->backward(grad_e_aev, result);
```
Comment (zasdfgbnm) [inline torchani/cuaev/cuaev.cpp:129]: I think we can just do
```suggestion
  Tensor grad_coord = cuaev_computer->backward(grad_e_aev, result_tensors);
```
and c++ will automatically convert tensor_list to Result?

File: torchani/cuaev/cuaev.cpp [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+  return grad_coord;
+}
+
+tensor_list CuaevDoubleAutograd::backward(AutogradContext* ctx, tensor_list grad_outputs) {
+  Tensor grad_force = grad_outputs[0];
+  torch::intrusive_ptr<CuaevComputer> cuaev_computer = ctx->saved_data["cuaev_computer"].toCustomClass<CuaevComputer>();
+  Result result(ctx->get_saved_variables());
+
+  Tensor grad_grad_aev = cuaev_computer->double_backward(grad_force, result);
```
Comment (zasdfgbnm) [inline torchani/cuaev/cuaev.cpp:143]: Same here
```suggestion
  Tensor grad_grad_aev = cuaev_computer->double_backward(grad_force, ctx->get_saved_variables());
```

File: torchani/cuaev/aev.h (hunk 6 above)
Comment (zasdfgbnm) [reply torchani/cuaev/aev.h:229]: It will speed up a tiny, probably not observable

File: torchani/cuaev/aev.h [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+      int64_t angular_length_aligned_,
+      int64_t ncenter_atoms_,
+      Tensor coordinates_t_,
+      Tensor species_t_);
+  Result(tensor_list tensors);
+  tensor_list to_list();
+  void release();
+  ~Result() {
+    this->release();
+  }
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.h:173]: I don't think they are needed.

File: torchani/cuaev/aev.h [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+      Tensor EtaR_t_,
+      Tensor ShfR_t_,
+      Tensor EtaA_t_,
+      Tensor Zeta_t_,
+      Tensor ShfA_t_,
+      Tensor ShfZ_t_,
+      int num_species_);
+};
+
+struct Result : torch::CustomClassHolder {
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.h:135]:
```suggestion
struct Result {
```

File: torchani/cuaev/aev.h [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+
+// cuda kernels
+Result cuaev_forward(const Tensor& coordinates_t, const Tensor& species_t, const AEVScalarParams& aev_params);
+Tensor cuaev_backward(const Tensor& grad_output, const AEVScalarParams& aev_params, const Result& result);
+Tensor cuaev_double_backward(const Tensor& grad_force, const AEVScalarParams& aev_params, const Result& result);
+
+// CuaevComputer
+// Only keep one copy of aev parameters and one copy of result for backward
+struct CuaevComputer : torch::CustomClassHolder {
+  AEVScalarParams aev_params;
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.h:184]: If `AEVScalarParams` is the only data field of `CuaevComputer`, should we merge these classes?

File: torchani/cuaev/aev.h [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+      Tensor tensor_centralAtom_,
+      Tensor tensor_numPairsPerCenterAtom_,
+      Tensor tensor_centerAtomStartIdx_,
+      int64_t maxnbrs_per_atom_aligned_,
+      int64_t angular_length_aligned_,
+      int64_t ncenter_atoms_,
+      Tensor coordinates_t_,
+      Tensor species_t_);
+  Result(tensor_list tensors);
+  tensor_list to_list();
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.h:169]: Is it better to make this an `operator tensor_list() {`?

File: torchani/cuaev/aev.h (hunk 13 above)
Comment (yueyericardo) [reply torchani/cuaev/aev.h:184]: better not, aev_params is needed to pass to cuda_forward, backward, double_backward functions.

Comment (zasdfgbnm) [reply torchani/cuaev/aev.h:184]: Could we just pass CuaevComputer to cuda_forward, backward, and double_backward?

**Review by zasdfgbnm — APPROVED** [review]

**Comment by yueyericardo** [comment]
Some notes about the memory issue are saved at: 
[Memory blow up when return a tensor from torch::CustomClassHolder by yueyericardo · Pull Request #1 · zasdfgbnm/test-autograd](https://github.com/zasdfgbnm/test-autograd/pull/1)

**Pull request merged**